# Artificial Neural Network for Customer’s Churn Prediction


Pooja Umathe


Churn’s prediction could be a great asset in the business strategy for retention applying before the exit of customers and We will create a real model with python, applied on a bank environment. This model will tell us if the customer is going or not to exit from the bank. Churn rate (sometimes called attrition rate), in its broadest sense, is a measure of the number of individuals or items moving out of a collective group over a specific period.

# What are neural networks?
Neural networks, commonly known as Artificial Neural Networks (ANN) are quite a simulation of human brain functionality in machine learning (ML) problems. ANNs shall be noted not as a solution for all the problems that arise, but would provide better results with many other techniques altogether for various ML tasks. Most common use of ANNs are clustering and classification, which can be used for regression tasks as well.

# Let's begin
So, In our dataset we would be dealing with Churn Modeling i.e. we would be writing a Artificial Neural Network to find out reasons as to why and which customers are actually leaving the bank and their dependencies on one another. This is a classification problem 0-1 classification(1 if customer Leaves and 0 if customer stays).

In [42]:
#Let’s import our python libraries and our dataset creating X for our independent variable and y for our dependent variable
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


In [43]:
# Importing the dataset
dataset = pd.read_csv('Bank_Churn.csv')
dataset.head(5)

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0       2       0.00              1          1               1   
1       1   83807.86              1          0               1   
2       8  159660.80              3          1               0   
3       1       0.00              2          0               0   
4       2  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [44]:
#Looking at the features we can see that row no.,name will have no relation with a customer with leaving the bank.
X = dataset.iloc[:, 3:13].values
y = dataset.iloc[:, 13].values

In [45]:
X

array([[619, 'France', 'Female', ..., 1, 1, 101348.88],
       [608, 'Spain', 'Female', ..., 0, 1, 112542.58],
       [502, 'France', 'Female', ..., 1, 0, 113931.57],
       ...,
       [709, 'France', 'Female', ..., 0, 1, 42085.58],
       [772, 'Germany', 'Male', ..., 1, 0, 92888.52],
       [792, 'France', 'Female', ..., 1, 0, 38190.78]], dtype=object)

Once we have quality data and selected the right target, we will prepare the data for the model. In general we do not need big data to create a model, but if our variables are significant, modelling with hundreds of data could work. We need always to test test our models to check if everything works correctly. Let’s say for our example to work with 10.000 rows dataset We will split our entire dataset in 2 parts. The bigger part, that will be 80% of data, will be used for the training of the model, while the remaining 20% will be used to test the model and have its accuracy.

But our model needs in input numerical data, so, we need to encode categorical data into numerical data. In This case we have Geography (France, Spain and Germany) and Gender (Male and Female). For Geography we will have 0,1,2 instead of France, Spain and Gemany and 0,1 instead of Gender.

In [46]:
# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X_1 = LabelEncoder()
X[:, 1] = labelencoder_X_1.fit_transform(X[:, 1])
labelencoder_X_2 = LabelEncoder()
X[:, 2] = labelencoder_X_2.fit_transform(X[:, 2])

In [47]:
# Now creating Dummy variables
onehotencoder = OneHotEncoder(categorical_features = [1])
X = onehotencoder.fit_transform(X).toarray()
X = X[:, 1:]

In [48]:
X

array([[0.0000000e+00, 0.0000000e+00, 6.1900000e+02, ..., 1.0000000e+00,
        1.0000000e+00, 1.0134888e+05],
       [0.0000000e+00, 1.0000000e+00, 6.0800000e+02, ..., 0.0000000e+00,
        1.0000000e+00, 1.1254258e+05],
       [0.0000000e+00, 0.0000000e+00, 5.0200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 1.1393157e+05],
       ...,
       [0.0000000e+00, 0.0000000e+00, 7.0900000e+02, ..., 0.0000000e+00,
        1.0000000e+00, 4.2085580e+04],
       [1.0000000e+00, 0.0000000e+00, 7.7200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 9.2888520e+04],
       [0.0000000e+00, 0.0000000e+00, 7.9200000e+02, ..., 1.0000000e+00,
        0.0000000e+00, 3.8190780e+04]])

In [49]:
# Splitting the dataset into the Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)


Before training it is important to apply feature scaling to data. It is an important step because it standardize the data and give them the same scale also for a faster computation.

In [50]:
#Feature scaling
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Now we can start to create our Artificial Neural Network and for doing this we will need to import the Keras library

Listing out the steps involved in training the ANN with Stochastic Gradient Descent:-

1)Randomly initialize the weights to small numbers close to 0(But not 0)
2)Input the 1st observation of your dataset in the Input Layer, each Feature in one Input Node
3)Forward-Propagation from Left to Right, the neurons are activated in a way that the impact of each neuron’s activation
is limited by the weights.Propagate the activations until getting the predicted result y.
4)Compare the predicted result with the actual result. Measure the generated error.
5)Back-Propagation: From Right to Left, Error is back propagated.Update the weights according to how much they are
responsible for the error.The Learning Rate tells us by how much such we update the weights.
6)Repeat Steps 1 to 5 and update the weights after each observation(Reinforcement Learning).
Or: Repeat Steps 1 to 5 but update the weights only after a batch of observations(Batch Learning) 
7)When the whole training set is passed through the ANN.That completes an Epoch. Redo more Epochs.

In [51]:
# Importing the Keras libraries and packages
import keras
from keras.models import Sequential
from keras.layers import Dense

We need now to define the number of hidden layers, the number of nodes for each hidden layers. We know that our input layers will have as many nodes as our independent variables (in our example we have 11 independent variables) and as output layers we will have our y, so 1 dependent variable.

In [52]:
# Initialising the ANN
classifier = Sequential()

We will use the Dense function A tip for the definition of the number of nodes of hidden layers, based on experiments, is to choose the average between the input and the output layers. In this case we will have 6 nodes (11+1)/2.

In [53]:
# Adding the input layer and the first hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu', input_dim = 11))

# Adding the second hidden layer
classifier.add(Dense(output_dim = 6, init = 'uniform', activation = 'relu'))

# Adding the output layer
classifier.add(Dense(output_dim = 1, init = 'uniform', activation = 'sigmoid'))

# Compiling the ANN
classifier.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

C:\Users\Rahul\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", input_dim=11, units=6, kernel_initializer="uniform")`
  
C:\Users\Rahul\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="relu", units=6, kernel_initializer="uniform")`
  """
C:\Users\Rahul\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(activation="sigmoid", units=1, kernel_initializer="uniform")`
  


In [54]:
# Fitting the ANN to the Training set
classifier.fit(X_train, y_train, batch_size = 10, nb_epoch = 100)

C:\Users\Rahul\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  


Epoch 1/100
8000/8000 [==============================] - 1s 141us/step - loss: 0.4862 - acc: 0.7956
Epoch 2/100
8000/8000 [==============================] - 1s 100us/step - loss: 0.4276 - acc: 0.7960
Epoch 3/100
8000/8000 [==============================] - 1s 108us/step - loss: 0.4220 - acc: 0.7960
Epoch 4/100
8000/8000 [==============================] - 1s 108us/step - loss: 0.4183 - acc: 0.8222
Epoch 5/100
8000/8000 [==============================] - 1s 93us/step - loss: 0.4162 - acc: 0.8255
Epoch 6/100
8000/8000 [==============================] - 1s 93us/step - loss: 0.4137 - acc: 0.8287
Epoch 7/100
8000/8000 [==============================] - 1s 92us/step - loss: 0.4128 - acc: 0.8309
Epoch 8/100
8000/8000 [==============================] - 1s 95us/step - loss: 0.4115 - acc: 0.8319
Epoch 9/100
8000/8000 [==============================] - 1s 94us/step - loss: 0.4104 - acc: 0.8340
Epoch 10/100
8000/8000 [==============================] - 1s 94us/step - loss: 0.4095 - acc: 0.8320
Epoch

In [55]:
# Making the predictions and evaluating the model
y_pred = classifier.predict(X_test)
y_pred = (y_pred > 0.5)


In [56]:
# Creating the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

In [57]:
cm

array([[1547,   48],
       [ 261,  144]], dtype=int64)

In [58]:
Accuracy = (1547+144)/(48+261+1547+144)
Accuracy

0.8455

The Accuracy of the Neural Network Model is 84.5% which is decent enough.